### PR TITLE
Features/unet_middle_block

### PR DESCRIPTION
**Pull request by zassou65535**
## 概要

Diffusionで必要となる、Unetの構成要素の1つ`UnetMiddleBlock`の追加

## Submit前の確認項目

- [x] タイトルは一目でわかるようにし、説明文は PR を簡潔に説明するようにしましたか?
- [x] あなたの PR が、異なる変更を束ねたものではなく、ひとつのことだけを行うものであることを確認しましたか?
- [x] この PR で導入されたすべての変更点をリストアップしましたか?
- [x] PR を`make run`コマンドでローカルでテストしましたか?

## 補足

`commons.py`内のクラス`AttentionBlock`に関して、実装の簡素化のために元実装でうまくいくか確認してから今後`nn.MultiheadAttention`で置き換える予定
実装が多少汚くても今後丸ごと削除予定なので問題ないと思われる

### CODE REVIEW

File: ami/models/components/unet/commons.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            in_channels (int):
+                Input num of channels.
+            emb_channels (int):
+                Input embedding num of channels.
+            out_channels (int):
+                Output num of channels.
+        """
+        super().__init__()
+        self.in_layers = nn.Sequential(
+            nn.GroupNorm(num_groups=32, num_channels=in_channels),
```
Comment (Geson-anko) [inline ami/models/components/unet/commons.py:30]: GroupNormの`num_groups`で`num_channels`は必ず割り切れないといけないから、コンストラクタの引数に`num_norm_groups=32`を追加してそれを指定する形にしよう。
```suggestion
            nn.GroupNorm(num_groups=num_norm_groups, num_channels=in_channels),
```

File: ami/models/components/unet/commons.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        h = self.in_layers(x)
+        emb_out = self.emb_layers(emb).type(h.dtype)
+        while len(emb_out.shape) < len(h.shape):
+            emb_out = emb_out[..., None]
+        # FiLM
+        out_norm, out_rest = self.out_layers[0], self.out_layers[1:]
+        scale, shift = torch.chunk(emb_out, 2, dim=1)
+        h = out_norm(h) * (1 + scale) + shift
+        h = out_rest(h)
+        return self.skip_connection(x) + h
```
Comment (Geson-anko) [inline ami/models/components/unet/commons.py:78]: nn.Sequentialでout_layersを定義しているが、indexでアクセスしたlayerがnorm layerであるとは限らない。
`self.out_norm=GroupNorm(...)`, `self.out_rest=nn.Sequential(...)`として別個に定義して行うべき。
```suggestion
        scale, shift = torch.chunk(emb_out, 2, dim=1)
        h = self.out_norm(h) * (1 + scale) + shift
        h = self.out_rest(h)
        return self.skip_connection(x) + h
```

File: ami/models/components/unet/commons.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                Output features.
+                (shape: [batch_size, out_channels, height, width])
+        """
+        h = self.in_layers(x)
+        emb_out = self.emb_layers(emb).type(h.dtype)
+        while len(emb_out.shape) < len(h.shape):
+            emb_out = emb_out[..., None]
+        # FiLM
+        out_norm, out_rest = self.out_layers[0], self.out_layers[1:]
+        scale, shift = torch.chunk(emb_out, 2, dim=1)
```
Comment (Geson-anko) [inline ami/models/components/unet/commons.py:75]: 最終次元に対して適用する意図
```suggestion
        scale, shift = torch.chunk(emb_out, 2, dim=-1)
```

File: ami/models/components/unet/middle_block.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                (shape: [batch_size, in_and_out_channels, height, width])
+            timestep_emb (torch.Tensor):
+                Embedded timesteps.
+                (shape: [batch_size, timestep_embed_dim])
+        Returns:
+            torch.Tensor
+                Output features.
+                (shape: [batch_size, in_and_out_channels, height, width])
+        """
+        assert len(self.resnet_blocks) == len(self.attentions) + 1
```
Comment (Geson-anko) [inline ami/models/components/unet/middle_block.py:58]: これはコンストラクタの定義から自明。もしassertかけるとしてもコンストラクタでやる。
```suggestion
```

File: ami/models/components/unet/middle_block.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            [
+                ResBlock(
+                    in_channels=in_and_out_channels, emb_channels=timestep_embed_dim, out_channels=in_and_out_channels
+                ),
+                ResBlock(
+                    in_channels=in_and_out_channels, emb_channels=timestep_embed_dim, out_channels=in_and_out_channels
+                ),
+            ]
+        )
+        self.attentions = nn.ModuleList([AttentionBlock(in_and_out_channels, num_heads=num_heads)])
```
Comment (Geson-anko) [inline ami/models/components/unet/middle_block.py:41]: わざわざModuleListにする必要がない。
```suggestion
        self.resblock_in = ResBlock(
                    in_channels=in_and_out_channels, emb_channels=timestep_embed_dim, out_channels=in_and_out_channels
                )
        self.resblock_out = ResBlock(
                    in_channels=in_and_out_channels, emb_channels=timestep_embed_dim, out_channels=in_and_out_channels
                )
        self.attention = AttentionBlock(in_and_out_channels, num_heads=num_heads)]
```

File: ami/models/components/unet/middle_block.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        Returns:
+            torch.Tensor
+                Output features.
+                (shape: [batch_size, in_and_out_channels, height, width])
+        """
+        assert len(self.resnet_blocks) == len(self.attentions) + 1
+        for resnet, attention in zip(self.resnet_blocks[:-1], self.attentions):
+            x = resnet(x, timestep_emb)
+            x = attention(x)
+        x = self.resnet_blocks[-1](x, timestep_emb)
```
Comment (Geson-anko) [inline ami/models/components/unet/middle_block.py:62]: 上の修正に基づいて。
```suggestion
    x = self.resblock_in(x, timestep_emb)
    x = self.attention(x)
    x = self.resblock_out(x, timestep_emb)
```

File: ami/models/components/unet/commons.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+    def __init__(
+        self,
+        channels: int,
+        num_heads: int,
+    ) -> None:
+        super().__init__()
+        self.channels = channels
+        self.num_heads = num_heads
+        self.norm = nn.GroupNorm(num_groups=32, num_channels=channels)
```
Comment (Geson-anko) [inline ami/models/components/unet/commons.py:100]: ここもコンストラクタで指定できるようにしよう。
```suggestion
        self.norm = nn.GroupNorm(num_groups=num_groups, num_channels=channels)
```

**Review by Geson-anko — CHANGES_REQUESTED** [review]
@zassou65535 実装ありがとう。修正お願いします。

File: ami/models/components/unet/middle_block.py (hunk 4 above)
Comment (zassou65535) [reply ami/models/components/unet/middle_block.py]: https://github.com/MLShukai/ami-vconf24/pull/67/commits/624c644b69f32cae80e7f578d695d08fa885fc69 で修正

File: ami/models/components/unet/commons.py (hunk 3 above)
Comment (zassou65535) [reply ami/models/components/unet/commons.py:85]: emb_outは` torch.Size([batch_size, emb_channels, 1, 1])`なはずだから、`dim=1`じゃないと分割に失敗する

File: ami/models/components/unet/commons.py (hunk 2 above)
Comment (zassou65535) [reply ami/models/components/unet/commons.py]: https://github.com/MLShukai/ami-vconf24/pull/67/commits/be8afb6edd06fb93d6b03287579a5b30ccc01be0 で変更

File: ami/models/components/unet/commons.py (hunk 3 above)
Comment (Geson-anko) [reply ami/models/components/unet/commons.py:87]: OK

**Review by Geson-anko — APPROVED** [review]
@zassou65535 Approveします
